directembedding/src/main/scala/ch/epfl/directembedding/transformers/EmbeddingTransformation.scala

```diff
@@ -0,0 +1,258 @@
+     *   self(q"new Foo()") == Some(q"Foo")
+     * }}}
+     */
+    private def self(lhs: Tree): Option[Tree] = lhs match {
+      case Select(x, _) if isNotModule(x) => Some(x)
+      case TypeApply(Select(x, _), _) if isNotModule(x) => Some(x)
+      case _ => None
+    }
+
+    /**
+     * Get types that are embedded inside the left-hand side of the application.
+     *
+     * Example: {{{
+     *   hiddenTypeArgs(q"new TArgClassExample[Int]") == List(tq"Int")
+     * }}}
+     */
+    private def hiddenTypeArgs(lhs: Tree): List[Tree] = lhs match {
+      case Select(x, _)        => x.tpe.typeArgs.map(TypeTree(_))
+      case TypeApply(_, targs) => targs
+      case _                   => Nil
+    }
+
+    private def symbol(lhs: Tree): Symbol = lhs match {
+      case Select(New(newBody), _)     => newBody.symbol
+      case field: Select               => invokedSymbol(field)
+      case TypeApply(field: Select, _) => invokedSymbol(field)
+      case _                           => lhs.symbol
+    }
+
+    private def selfType(tree: Tree): Type = {
+      val owner = tree.symbol.owner
+      if (owner.isType) owner.asType.toType
+      else c.abort(tree.pos, s"Unable to resolve type for $tree")
+    }
+
+    /**
+     * Used for finding annotations for predefined/third-party definitions by looking
+     * up into the type map that points to the fallback type.
+     */
+    private def fallbackAnnotation(treeSymbol: Symbol, tree: Tree): Iterable[Annotation] = {
+      val sym = for {
+        fallbackType <- typeMap.get(selfType(tree).typeSymbol.fullName)
+        fallbackSymbol <- {
+          log(s"Falling back to $fallbackType")
+          findMatchingSymbol(treeSymbol, fallbackType)
+        }
+      } yield fallbackSymbol
+      sym.toList.flatMap(_.annotations)
+    }
+
+    /**
+     * Strip away sugar from partially applied functions in annotations.
+     *
+     * Example:
+     *   {{{@reifyAs(method _)}}}
+     * is eta expanded into:
+     *   {{{(args: Args*) => method(args)}}}
+     * {{{
+     *   stripEtaExpansion(q"method ") == q"method"
+     * }}}
+     */
+    private def stripEtaExpansion(tree: Tree): Option[Tree] = {
+      tree match {
+        case Block(_, Function(lhs, rhs)) => rhs.collect {
+          case Apply(s: Select, _)     => s
+          case TypeApply(s: Select, _) => s
+        }.headOption
+        case _ => Some(tree)
+      }
+    }
+
+    private def annotationBody(annotation: Annotation): Tree =
+      annotation.tree.children.tail.headOption.flatMap(stripEtaExpansion).getOrElse {
+        c.abort(annotation.tree.pos, s"Missing argument for $annotation")
+      }
+
+    private var indent: Int = 0
+
+    override def transform(tree: Tree): Tree = {
+      logIndented(s"transform(): $tree", indent)
+      logTree(tree, logLevel + 1)
+      indent += 2
+      val result = tree match {
+        // f(a)(b) => f(a, b)
+        case a @ Apply(Apply(_, _), _) if flattenCurriedFunctions =>
```

Comment:
Will add issue to make this automatic. I don't remember though the exact details of how that was gonna work.